### PR TITLE
#70 Use blocks defined in the middle of template

### PR DESCRIPTION
**Pull request by paweltarkowski**

### CODE REVIEW

File: src/main/kotlin/co/nums/intellij/aem/htl/inspections/HtmlMiddleTemplateBlocksVariablesAnnotator.kt [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import com.intellij.psi.xml.XmlFile
+import com.intellij.psi.xml.XmlTag
+import org.jetbrains.annotations.NotNull
+
+class HtmlMiddleTemplateBlocksVariablesAnnotator : XmlSuppressableInspectionTool() {
+
+    private val MESSAGE = "Use data-sly-use variable statements only on root elements"
+
+    override fun getDisplayName(): String {
+        return "Middle template blocks variables annotator"
```
Comment (karollewandowski) [inline src/main/kotlin/co/nums/intellij/aem/htl/inspections/HtmlMiddleTemplateBlocksVariablesAnnotator.kt]: Please use single-expression functions in such cases:
```java
override fun getDisplayName() = "Middle template blocks variables annotator"
```

File: src/main/resources/META-INF/plugin.xml [hunk 2]
```diff
@@ -102,8 +102,13 @@
 				implementationClass="co.nums.intellij.aem.htl.editor.braces.HtlBraceMatcher"/>
 
 		<annotator
-				language="HTL"
```
Comment (karollewandowski) [inline src/main/resources/META-INF/plugin.xml]: Accidentally reformatted?

File: src/main/kotlin/co/nums/intellij/aem/htl/psi/extensions/XmlElementVisitorExtension.kt [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package co.nums.intellij.aem.htl.psi.extensions
```
Comment (karollewandowski) [inline src/main/kotlin/co/nums/intellij/aem/htl/psi/extensions/XmlElementVisitorExtension.kt]: `extension` packages are for Kotlin extension functions.

File: src/main/kotlin/co/nums/intellij/aem/htl/inspections/HtmlMiddleTemplateBlocksVariablesAnnotator.kt [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        return "Select non root data-sly-use statements variables"
+    }
+
+
+    override fun isEnabledByDefault(): Boolean {
+        return true
+    }
+
+    fun checkAttribute(attribute: XmlAttribute, holder: ProblemsHolder, firstElement: XmlTag?) {
+        if (attribute.isHtlVariableBlock() && attribute.name.startsWith("data-sly-use") && (firstElement != null && firstElement != attribute.parent)) {
```
Comment (karollewandowski) [inline src/main/kotlin/co/nums/intellij/aem/htl/inspections/HtmlMiddleTemplateBlocksVariablesAnnotator.kt]: I think we can create new extension function `XmlAttribute.isHtlUseBlock()`, so there won't be need for `attribute.name.startsWith("data-sly-use")` and abstractions levels won't be mixed.

File: src/main/kotlin/co/nums/intellij/aem/htl/inspections/HtmlMiddleTemplateBlocksVariablesAnnotator.kt (hunk 1 above)
Comment (paweltarkowski) [inline src/main/kotlin/co/nums/intellij/aem/htl/inspections/HtmlMiddleTemplateBlocksVariablesAnnotator.kt]: I adjust it

File: src/main/resources/META-INF/plugin.xml (hunk 2 above)
Comment (paweltarkowski) [inline src/main/resources/META-INF/plugin.xml]: yes

File: src/main/kotlin/co/nums/intellij/aem/htl/psi/extensions/XmlElementVisitorExtension.kt (hunk 3 above)
Comment (paweltarkowski) [inline src/main/kotlin/co/nums/intellij/aem/htl/psi/extensions/XmlElementVisitorExtension.kt]: I rename class into XmlElementVisitorImpl.
In which package you propose to move this class ?

File: src/main/kotlin/co/nums/intellij/aem/htl/inspections/HtmlMiddleTemplateBlocksVariablesAnnotator.kt (hunk 4 above)
Comment (paweltarkowski) [inline src/main/kotlin/co/nums/intellij/aem/htl/inspections/HtmlMiddleTemplateBlocksVariablesAnnotator.kt]: I agree :)

**Comment by karollewandowski** [comment]
I'm closing it and will continue review in https://github.com/karollewandowski/aem-intellij-plugin/pull/80

**Pull request closed without merging**